src/main/java/com/karankumar/bookproject/backend/service/ImportService.java

```diff
@@ -0,0 +1,129 @@
+        return savedBooks;
+    }
+
+    private List<Book> toBooks(Collection<? extends GoodreadsBookImport> goodreadsBookImports) {
+        return goodreadsBookImports.stream()
+                                   .map(this::toBook)
+                                   .filter(Optional::isPresent)
+                                   .map(Optional::get)
+                                   .collect(Collectors.toList());
+    }
+
+    private Optional<Book> toBook(GoodreadsBookImport goodreadsBookImport) {
+        Optional<Author> author = toAuthor(goodreadsBookImport.getAuthor());
+        if (author.isEmpty()) {
+            LOGGER.severe("Author is null");
+            return Optional.empty();
+        }
+
+        Optional<PredefinedShelf> predefinedShelf =
+                toPredefinedShelf(goodreadsBookImport.getBookshelves(),
+                        goodreadsBookImport.getDateRead());
+        if (predefinedShelf.isEmpty()) {
+            LOGGER.severe("Predefined shelf is null");
+            return Optional.empty();
+        }
+
+        Book book = new Book(goodreadsBookImport.getTitle(), author.get(), predefinedShelf.get());
+
+        Optional<CustomShelf> customShelf = toCustomShelf(goodreadsBookImport.getBookshelves());
+        customShelf.ifPresent(book::setCustomShelf);
+
+        Optional<RatingScale> ratingScale = toRatingScale(goodreadsBookImport.getRating(), 2);
+        ratingScale.ifPresent(book::setRating);
+
+        return Optional.of(book);
+    }
+
+    private Optional<Author> toAuthor(String name) {
+        if (StringUtils.isBlank(name)) {
+            return Optional.empty();
+        }
+        String[] authorNames = name.split(" ");
+        return Optional
+                .of(new Author(authorNames[0], authorNames.length > 1 ? authorNames[1] : null));
+    }
+
+    private Optional<PredefinedShelf> toPredefinedShelf(String shelves, LocalDate dateRead) {
+        if (Objects.nonNull(dateRead)) {
+            return Optional.of(predefinedShelfService.findToReadShelf());
+        }
+        if (StringUtils.isBlank(shelves)) {
+            return Optional.empty();
+        }
+        String[] shelvesArray = shelves.split(",");
+        return Arrays.stream(shelvesArray)
+                     .filter(PredefinedShelfUtils::isPredefinedShelf)
+                     .findFirst()
+                     .map(predefinedShelfService::getPredefinedShelfByNameAsString);
+    }
+
+    private Optional<CustomShelf> toCustomShelf(String shelves) {
+        if (StringUtils.isBlank(shelves)) {
+            return Optional.empty();
+        }
+        String[] shelvesArray = shelves.split(",");
+        return Arrays.stream(shelvesArray)
+                     .filter(Predicate.not(PredefinedShelfUtils::isPredefinedShelf))
+                     .findFirst()
+                     .map(customShelfService::findOrCreate);
+    }
+
+    private Optional<RatingScale> toRatingScale(Double ratingValue, double scaleFactor) {
```

Comment:
As the scale factor is always 2, could we make this a constant, instead?

Reply:
That is true for goodreads. But the service and inturn the method is source agnostic, that is why I have taken a parameter here as the scale factor might be different for various third party data stores.

Reply:
That makes sense. Could you create a variable and pass it into the method to make it a little clearer that the 2 is because our scale is 2x Goodreads's?

Reply:
Sure. I'll do that.